clang/lib/Sema/SemaCodeComplete.cpp

```diff
@@ -6292,11 +6298,16 @@ SemaCodeCompletion::ProduceCallSignatureHelp(Expr *Fn, ArrayRef<Expr *> Args,
         SmallVector<Expr *, 12> ArgExprs(1, NakedFn);
         ArgExprs.append(ArgsWithoutDependentTypes.begin(),
                         ArgsWithoutDependentTypes.end());
+        auto *const LambdaName =
+            DC->isLambda() ? cast<VarDecl>(NakedFn->getReferencedDeclOfCallee())
```

Comment:
I vaguely recall this function not being particularly reliable for lambdas.  Also, are we sure this is always a VarDecl?  I thought there are cases where this is the return of a function (that is, a function returning a lambda, now being called) where this would be the function instead.  So at minimum this would need a test for that (and this might have to be a dyn_cast).

Reply:
So in this context, I presume `NakedFn` is a `DeclRefExpr` that refers to the `VarDecl` of the declaration? Can we assert it?
(It would be great if you can add tests from Erich above)

Reply:
Also, if my theory is correct, this could be converted to
```
dyn_cast_if_present<VarDecl>(NakedFn->getReferencedDeclOfCallee())
```

Reply:
I don't think you could just do `cast_if_present`, I'm having troubles coming up with times where that wouldn't be a `VarDecl`, but I very much feel like it must be able to.